Setoption147 (MQTT) Disable publish SSerialReceived MQTT messages. If disabled, you must use event trigger rules instead.

ortegafernando opened the pull request:
## Description:

This option helps when we have a device that send by its own serial data every X seconds.

## Checklist:
  - [X] The pull request is done against the latest development branch
  - [X] Only relevant files were touched
  - [X] Only one feature/fix was added per PR and the code change compiles without warnings
  - [X] The code change is tested and works with Tasmota core ESP8266 V.2.7.4.9
  - [X] The code change is tested and works with Tasmota core ESP32 V.2.0.5
  - [X] I accept the [CLA](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md#contributor-license-agreement-cla).

_NOTE: The code change must pass CI tests. **Your PR cannot be merged unless tests pass**_

Add SetOption147  (MQTT) Enable publish SSerialReceived MQTT messages. If disabled, use event rules instead.

If it is disabled, tasmota will not publish an MQTT message every time a SSerialReceived event fires.

Use rules to control how and when publish these messages.

DEfault value is ENABLED so anyone is affected if they don't change this 147 option value.

ortegafernando committed:
Add SetOption147

Add SetOption147  (MQTT) Enable publish SSerialReceived MQTT messages. If disabled, use event rules instead.
If it is disabled, tasmota will not publish an MQTT message every time a SSerialReceived event fires. Use rules to control how and when publish these messages.

ortegafernando committed:
Add SetOption147

Add SetOption147  (MQTT) Enable publish SSerialReceived MQTT messages. If disabled, use event rules instead.
If it is disabled, tasmota will not publish an MQTT message every time a SSerialReceived event fires. Use rules to control how and when publish these messages.

ortegafernando committed:
Add SetOption147

Add SetOption147  (MQTT) Enable publish SSerialReceived MQTT messages. If disabled, use event rules instead.
If it is disabled, tasmota will not publish an MQTT message every time a SSerialReceived event fires. Use rules to control how and when publish these messages.

ortegafernando committed:
Add SetOption157

Add SetOption147  (MQTT) Enable publish SSerialReceived MQTT messages. If disabled, use event rules instead.
If it is disabled, tasmota will not publish an MQTT message every time a SSerialReceived event fires. Use rules to control how and when publish these messages.

ortegafernando committed:
Add SetOption147

Add SetOption147  (MQTT) Enable publish SSerialReceived MQTT messages. If disabled, use event rules instead.
If it is disabled, tasmota will not publish an MQTT message every time a SSerialReceived event fires. Use rules to control how and when publish these messages.

ortegafernando committed:
Add SetOption147

Add SetOption147  (MQTT) Enable publish SSerialReceived MQTT messages. If disabled, use event rules instead.
If it is disabled, tasmota will not publish an MQTT message every time a SSerialReceived event fires. Use rules to control how and when publish these messages.

barbudor commented:
This is not acceptable as it is: it is a double breaking cjange

1) Please revert the signification so that enabling the option disable the publication
2) the So should only disable mqtt, not rules
It shouldn't be one or the other

ortegafernando commented:
> This is not acceptable as it is: it is a double breaking cjange
> 
> 1. Please revert the signification so that enabling the option disable the publication
> 2. the So should only disable mqtt, not rules
>    It shouldn't be one or the other

So. 
1. You think that this will be better, won't it?
```
uint32_t mqtt_disable_sserialrec : 1;     // bit 1  (v12.x.x.x) - SetOption147 - (MQTT) Disable publish SSerialReceived MQTT messages, you must use event trigger rules instead.
```
```
#ifndef MQTT_SSERIALRECEIVED
#define MQTT_DISABLE_SSERIALRECEIVED        0          // 1 = Disable sserialreceived mqtt messages, 0 = Enable sserialreceived mqtt messages (default)
#endif
```
And of course, i will have to change "if" condition.

2. I dont exactly understand you, this will be the if condition:
```
    if (Settings->flag6.mqtt_disable_sserialrec ) {  // SetOption147  If Enabled, Tasmota will not publish SSerialReceived MQTT messages, but it will proccess event trigger rules
       XdrvRulesProcess(0);
    } else {
      MqttPublishPrefixTopicRulesProcess_P(RESULT_OR_TELE, PSTR(D_JSON_SSERIALRECEIVED));
    }
```

is it now clear? (sorry I am not a native english so I used some wrong words. I think now it is better, thanks for your help)

barbudor commented:
Hi
My first message was done on phone in the bus so I didn't spend much words in explanations 😄 

When a new feature is added, it shouldn't change the behavior of Tasmota with regards to other users
The current behavior is on SSerial receive:
- Publish MQTT
- AND ProcessRule
So, considering that the SO bit is by default 0 on every Tasmota devices, this behavior shouldn't be changed for any user upgrading after the addition of the feature.
This means:
- the SO bit should be defined as a 'disable feature'
- when it is 0 (default) the behavior shouldn't be change and both MQTT Publication and Rule process should happen

In the bus I didn't had the time to check that `MqttPublishPrefixTopicRulesProcess_P()` was also calling the `XdrvRulesProcess()`
So indeed your code was correct and I apologies for that. I thought you were totally disabling the Rules.

However, I do prefer the 2nd form which makes it very clear what the SO bit is doing

I think you just have to replace `NOT` by `!` 😄 

Please note that this is my advice on coding with regards to Tasmota standards.
At the end it is Theo who will decide if this is worth to be merged
I suggest that you edit your description to add the link to the Discussion for him to understand the need.

ortegafernando commented:
> Hi My first message was done on phone in the bus so I didn't spend much words in explanations 😄
> 
> When a new feature is added, it shouldn't change the behavior of Tasmota with regards to other users The current behavior is on SSerial receive:
> 
> * Publish MQTT
> * AND ProcessRule
>   So, considering that the SO bit is by default 0 on every Tasmota devices, this behavior shouldn't be changed for any user upgrading after the addition of the feature.
>   This means:
> * the SO bit should be defined as a 'disable feature'
> * when it is 0 (default) the behavior shouldn't be change and both MQTT Publication and Rule process should happen
> 
> In the bus I didn't had the time to check that `MqttPublishPrefixTopicRulesProcess_P()` was also calling the `XdrvRulesProcess()` So indeed your code was correct and I apologies for that. I thought you were totally disabling the Rules.
> 
> However, I do prefer the 2nd form which makes it very clear what the SO bit is doing
> 
> I think you just have to replace `NOT` by `!` 😄
> 
> Please note that this is my advice on coding with regards to Tasmota standards. At the end it is Theo who will decide if this is worth to be merged I suggest that you edit your description to add the link to the Discussion for him to understand the need.

Ok, thanks a lot. I have laught a lot about the bus 😄.

I will wait Teo about which option he thinks it is better. This option is not ram or flash high consuming code so I hope Teo will implement it. (it is only few bytes)

ortegafernando committed:
Setoption147

Setoption147 (MQTT) Disable publish SSerialReceived MQTT messages. If disabled, you must use event trigger rules instead.

If it is activated, tasmota will not publish a MQTT message every time a SSerialReceived event fires. Use rules to control "what, how and when" publish these messages.

ortegafernando committed:
Setoption147

Setoption147 (MQTT) Disable publish SSerialReceived MQTT messages. If disabled, you must use event trigger rules instead.

If it is activated, tasmota will not publish a MQTT message every time a SSerialReceived event fires. Use rules to control "what, how and when" publish these messages.

ortegafernando committed:
Setoption147

Setoption147 (MQTT) Disable publish SSerialReceived MQTT messages. If disabled, you must use event trigger rules instead.

If it is activated, tasmota will not publish a MQTT message every time a SSerialReceived event fires. Use rules to control "what, how and when" publish these messages.

ortegafernando committed:
SetOption147

Setoption147 (MQTT) Disable publish SSerialReceived MQTT messages. If disabled, you must use event trigger rules instead.

If it is activated, tasmota will not publish a MQTT message every time a SSerialReceived event fires. Use rules to control "what, how and when" publish these messages.

ortegafernando commented:
Hi, as @Jason2866  has marked this PR ready for review I have updated files with the @barbudor tips.
May be the only thing to think about is the "if condition". I have written as I think is better, but there are two possibilities.

About sizes (using gitpod.io):

Without this PR:
```
RAM:   [=====     ]  50.6% (used 41424 bytes from 81920 bytes)
Flash: [======    ]  62.2% (used 636840 bytes from 1023984 bytes)
Compression reduced firmware size by 71% (was 640992 bytes, now 457374 bytes)
```

With this PR:
```
RAM:   [=====     ]  50.6% (used 41424 bytes from 81920 bytes)
Flash: [======    ]  62.2% (used 636856 bytes from 1023984 bytes)
Compression reduced firmware size by 71% (was 641008 bytes, now 457410 bytes)
```

16 bytes of flash.

s-hadinger commented on tasmota/include/tasmota_types.h:
```diff
@@ -180,7 +180,7 @@ typedef union {                            // Restricted by MISRA-C Rule 18.4 bu
   struct {                                 // SetOption146 .. SetOption177
     uint32_t use_esp32_temperature : 1;    // bit 0  (v12.1.1.1) - SetOption146 - (ESP32) Show ESP32 internal temperature sensor
-    uint32_t spare01 : 1;                  // bit 1
+    uint32_t mqtt_disable_sserialrec : 1;  // bit 1  (v12.x.x.x) - SetOption147 - (MQTT) Disable publish SSerialReceived MQTT messages, you must use event trigger rules instead.
```
Can you please add the current version of Tasmota?

ortegafernando replied:
Ok, I didn't know if I need to put the actual version or the next version. Thanks.

s-hadinger reviewed (commented):
Thanks for your PR. I have submitted a request for a few additions.

ortegafernando committed:
Add SetOption147

SetOption147 (MQTT) Disable publish SSerialReceived MQTT messages. If disabled, you must use event trigger rules instead.

If it is activated, tasmota will not publish a MQTT message every time a SSerialReceived event fires. Use rules to control "what, how and when" publish these messages.

ortegafernando commented on tasmota/tasmota_support/settings.ino:
```diff
@@ -838,6 +838,7 @@ void SettingsDefaultSet2(void) {
   SOBitfield3  flag3 = { 0 };
   SOBitfield4  flag4 = { 0 };
   SOBitfield5  flag5 = { 0 };
+  SOBitfield6  flag6 = { 0 };
```
Done, thanks.

ortegafernando committed:
Add SetOption147

SetOption147 (MQTT) Disable publish SSerialReceived MQTT messages. If disabled, you must use event trigger rules instead.

If it is activated, tasmota will not publish a MQTT message every time a SSerialReceived event fires. Use rules to control "what, how and when" publish these messages.

ortegafernando committed:
Add SetOption147

SetOption147 (MQTT) Disable publish SSerialReceived MQTT messages. If disabled, you must use event trigger rules instead.

If it is activated, tasmota will not publish a MQTT message every time a SSerialReceived event fires. Use rules to control "what, how and when" publish these messages.

ortegafernando commented on tasmota/include/tasmota_globals.h:
```diff
@@ -283,6 +283,9 @@ String EthernetMacAddress(void);
 #ifndef MQTT_CLEAN_SESSION
 #define MQTT_CLEAN_SESSION          1          // 0 = No clean session, 1 = Clean session (default)
 #endif
+#ifndef MQTT_DISABLE_SSERIALRECEIVED        
```
Thanks, Done.

ortegafernando commented:
Hi, I think I have solved all reviews. I am not sure about Tasmota version. 

Do I will need to update the wiki after this PR is accepted or is it done by any admin? I don't mind, of course.

And what about "changelog"?

Thanks.

s-hadinger reviewed (approved):
This is good for me, thanks. Let's wait for Theo to come back and merge

ortegafernando committed:
Add SetOption147

SetOption147 (MQTT) Disable publish SSerialReceived MQTT messages. If disabled, you must use event trigger rules instead.

If it is activated, tasmota will not publish a MQTT message every time a SSerialReceived event fires. Use rules to control "what, how and when" publish these messages.

ortegafernando commented:
Add PR for Wiki.

https://github.com/tasmota/docs/pull/1053
